### PR TITLE
Support for nil comparisons without Equatable conformance

### PR DESCRIPTION
**Pull request by MatthewTurk247**
This pull request implements the builder functions proposed in #2117.

### Testing:

See `PredicateTests/nilLiteral(equality:)` and `PredicateTests/nilLiteral(inequality:)`. Both exercise the new functions since `PredicateTestObject` does not conform to `Equatable`. For macro expansions, see `PredicateMacroLanguageTokenTests/nilLiterals()`.

### CODE REVIEW

**Comment by jmschonfeld** [comment]
@swift-ci please test

**Review by jmschonfeld — APPROVED** [review]

File: Sources/FoundationEssentials/Predicate/Expressions/Equality.swift [hunk 1]
```diff
@@ -40,6 +40,17 @@ extension PredicateExpressions {
     }
 }
 
+@available(FoundationPreview 6.4, *)
```
Comment (jmschonfeld) [inline Sources/FoundationEssentials/Predicate/Expressions/Equality.swift]:
```suggestion
@available(FoundationPreview 6.5, *)
```

File: Sources/FoundationEssentials/Predicate/Expressions/Inequality.swift [hunk 2]
```diff
@@ -40,6 +40,17 @@ extension PredicateExpressions {
     }
 }
 
+@available(FoundationPreview 6.4, *)
```
Comment (jmschonfeld) [inline Sources/FoundationEssentials/Predicate/Expressions/Inequality.swift]:
```suggestion
@available(FoundationPreview 6.5, *)
```

File: Sources/FoundationMacros/PredicateMacro.swift [hunk 3]
```diff
@@ -496,8 +496,29 @@ private class PredicateQueryRewriter: SyntaxRewriter, PredicateSyntaxRewriter {
...
+        if isNilLHS && isNilRHS {
+            let builder: String? = switch opSyntax.operator.text {
+            case "==": "build_Equal"
+            case "!=": "build_NotEqual"
+            default: nil
+            }
+            if let builder {
+                return """
+                \(raw: indent)PredicateExpressions.\(raw: moduleName())::\(raw: builder)(
+                \(raw: indent + indentWidth)lhs: PredicateExpressions.\(raw: moduleName())::build_NilLiteral() as PredicateExpressions.\(raw: moduleName())::NilLiteral<Swift.Bool>,
```
Comment (jmschonfeld) [inline Sources/FoundationMacros/PredicateMacro.swift:511]: Is this guaranteed to be constant-folded? In other words, it's not possible to end up with a dynamic cast here, right? This is definitely a scenario where I wish we could specify the generic type parameters of a function at the call site to be more explicit about what's going on here

File: Tests/FoundationMacrosTests/PredicateMacroLanguageTokenTests.swift [hunk 4]
```diff
@@ -393,7 +393,55 @@ private struct PredicateMacroLanguageTokenTests {
...
+            """
+            #Predicate<Object?> { input in
+                nil != input
+            }
+            """,
+            """
+            \(foundationModuleName).Predicate<Object?>({ input in
+                PredicateExpressions.\(foundationModuleName)::build_NotEqual(
+                    nilLiteral: PredicateExpressions.\(foundationModuleName)::build_NilLiteral(),
+                    rhs: PredicateExpressions.\(foundationModuleName)::build_Arg(input)
```
Comment (jmschonfeld) [inline Tests/FoundationMacrosTests/PredicateMacroLanguageTokenTests.swift:444]: Can you add a syntax transformation test for the `nil == nil` case as well?

Comment (MatthewTurk247) [reply Tests/FoundationMacrosTests/PredicateMacroLanguageTokenTests.swift:444]: Yes, will do. I thought I had added a `nil == nil` syntax test earlier but might’ve lost track of it.

File: Sources/FoundationMacros/PredicateMacro.swift (hunk 3 above)
Comment (MatthewTurk247) [reply Sources/FoundationMacros/PredicateMacro.swift:511]: From what I can tell, there’s no dynamic cast. The `as` is in the expansion solely to provide enough information for type inference to succeed. Shouldn’t be necessary to emit `checked_cast`, and locally, it does not appear in the SILGen for a builder closure:

```
sil private [ossa] @$s3fooAA10Foundation9PredicateVySiSg_QPGyFAB08StandardC10Expression_pSb6OutputAB0cE0PRts_XPAB0C11ExpressionsO8VariableVy_AEGXEfU_ : $@convention(thin) @substituted <each τ_0_0> (@pack_guaranteed Pack{repeat PredicateExpressions.Variable<each τ_0_0>}) -> @out any StandardPredicateExpression<Bool> for <Pack{Optional<Int>}> {
// %0 "$return_value"                             // user: %15
// %1                                             // user: %3
bb0(%0 : $*any StandardPredicateExpression<Bool>, %1 : $*Pack{PredicateExpressions.Variable<Optional<Int>>}):
  %2 = scalar_pack_index 0 of $Pack{PredicateExpressions.Variable<Optional<Int>>} // user: %3
  %3 = pack_element_get %2 of %1 as $*PredicateExpressions.Variable<Optional<Int>> // user: %4
  debug_value %3, let, name "input", argno 1, expr op_deref // id: %4
  %5 = metatype $@thin PredicateExpressions.Type  // user: %16
  %6 = alloc_stack $PredicateExpressions.NilLiteral<Bool> // users: %20, %19, %16, %9
  %7 = metatype $@thin PredicateExpressions.Type  // user: %9
  // function_ref static PredicateExpressions.build_NilLiteral<A>()
  %8 = function_ref @$s10Foundation20PredicateExpressionsO16build_NilLiteralAC0eF0Vy_xGylFZ : $@convention(method) <τ_0_0> (@thin PredicateExpressions.Type) -> @out PredicateExpressions.NilLiteral<τ_0_0> // user: %9
  %9 = apply %8<Bool>(%6, %7) : $@convention(method) <τ_0_0> (@thin PredicateExpressions.Type) -> @out PredicateExpressions.NilLiteral<τ_0_0>
  %10 = alloc_stack $PredicateExpressions.NilLiteral<Bool> // users: %18, %17, %16, %13
  %11 = metatype $@thin PredicateExpressions.Type // user: %13
  // function_ref static PredicateExpressions.build_NilLiteral<A>()
  %12 = function_ref @$s10Foundation20PredicateExpressionsO16build_NilLiteralAC0eF0Vy_xGylFZ : $@convention(method) <τ_0_0> (@thin PredicateExpressions.Type) -> @out PredicateExpressions.NilLiteral<τ_0_0> // user: %13
  %13 = apply %12<Bool>(%10, %11) : $@convention(method) <τ_0_0> (@thin PredicateExpressions.Type) -> @out PredicateExpressions.NilLiteral<τ_0_0>
  // function_ref static PredicateExpressions.build_Equal<A, B>(lhs:rhs:)
  %14 = function_ref @$s10Foundation20PredicateExpressionsO11build_Equal3lhs3rhsAC0E0Vy_xq_Gx_q_tAA0B10ExpressionRzAaJR_SQ6OutputRpzAKQy_ALRSr0_lFZ : $@convention(method) <τ_0_0, τ_0_1 where τ_0_0 : PredicateExpression, τ_0_1 : PredicateExpression, τ_0_0.Output : Equatable, τ_0_0.Output == τ_0_1.Output> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_1, @thin PredicateExpressions.Type) -> @out PredicateExpressions.Equal<τ_0_0, τ_0_1> // user: %16
  %15 = init_existential_addr %0, $PredicateExpressions.Equal<PredicateExpressions.NilLiteral<Bool>, PredicateExpressions.NilLiteral<Bool>> // user: %16
  %16 = apply %14<PredicateExpressions.NilLiteral<Bool>, PredicateExpressions.NilLiteral<Bool>>(%15, %6, %10, %5) : $@convention(method) <τ_0_0, τ_0_1 where τ_0_0 : PredicateExpression, τ_0_1 : PredicateExpression, τ_0_0.Output : Equatable, τ_0_0.Output == τ_0_1.Output> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_1, @thin PredicateExpressions.Type) -> @out PredicateExpressions.Equal<τ_0_0, τ_0_1>
  destroy_addr %10                                // id: %17
  dealloc_stack %10                               // id: %18
  destroy_addr %6                                 // id: %19
  dealloc_stack %6                                // id: %20
  %21 = tuple ()                                  // user: %22
  return %21                                      // id: %22
} // end sil function '$s3fooAA10Foundation9PredicateVySiSg_QPGyFAB08StandardC10Expression_pSb6OutputAB0cE0PRts_XPAB0C11ExpressionsO8VariableVy_AEGXEfU_’
```
